### PR TITLE
Includes complete point addition and tests

### LINKED ISSUE #62: ecc/p384: ScalarMult fails for k=6
The P384 `ScalarMult` implementation does not return the correct result for the scalar `k = 6`. For a reproducer see:

https://gist.github.com/mmcloughlin/823f698cd20cb299327bec705de0f014

One example result:

```
$ go run k6.go
got    x=0 y=0
expect x=78d0da4eeeb585f21e4de75c543807e8bd7d479a9b332f2cd5e4fbfc219e54d328ee02cce588805bf5ba9217fc52e1c8 y=723df279aaf436b7f088d60078cf26aef0410a51fbc1ec241532ebf321192749dfde358aef76e8874148c321761984d8
```

I discovered this because I noticed that the `ScalarMult` function does not implement the same algorithm as described in [Selecting Elliptic Curves for Cryptography: An Efficiency and Security Analysis](https://eprint.iacr.org/2014/130). It is very similar to Algorithm 1, but does not implement Step 19 which uses a complete addition formula for the final add.

Screen Shot 2019-09-01 at 4 15 10 PM

The value k=6 has the property that the final add of the algorithm will be `(-3)P + (-3)P`. Therefore the `add` function fails, since it does not support doubling.

### PR DESCRIPTION
**Pull request by armfazh**
Changes:
 - Adds complete point addition of [Renes/Costello/Batina](ia.cr/2015/1060)
 - Point addition in homogeneous coordinates.
 - Representation of point at infinity.
 -- On afine, (0,0) is the identity point.
 -- On Jacobian, (0:0:0) is not a valid point.
 -- On homogeneous, (0:0:0) is not a valid point.
 - Added a zero point initializer.

Fixes #62.

### CODE REVIEW

**Comment by armfazh** [comment]
_ready for review_

File: math/wnaf.go [hunk 1]
```diff
@@ -3,24 +3,33 @@ package math
 
 import "math/big"
 
-// SignedDigit obtains the signed-digit recoding of n such that n>0 and odd and
-// returns a list L such that n = sum( L[i]*2^(w-1) ) and L[i] are odd numbers
-// in the set {±1, ±3, ..., ±2^(w-1)-1}, where 1 < w < 32.
+// SignedDigit obtains the signed-digit recoding of n and returns a list L of
+// digits such that n = sum( L[i]*2^(w-1) ), and each L[i] is an odd number
```
Comment (wbl) [inline math/wnaf.go]: I think i belongs in the exponent somewhere.

**Comment by henrydcase** [comment]
2015/1060

File: math/wnaf.go (hunk 1 above)
Comment (armfazh) [reply math/wnaf.go]: good catch!

File: ecc/p384/point.go [hunk 2]
```diff
@@ -246,3 +268,95 @@ func (P *jacobianPoint) double() {
...
+func (P *jacobianPoint) toHomogeneous() *homogeneousPoint {
+	var hP homogeneousPoint
+	hP.y = P.y
+	fp384Mul(&hP.x, &P.x, &P.z)
+	fp384Sqr(&hP.z, &P.z)
+	fp384Mul(&hP.z, &hP.z, &P.z)
+	return &hP
+}
+
+// homogeneousPoint represents a point in homogeneous coordinates.
```
[the remark this answers is not in the document]
Comment (armfazh) [reply ecc/p384/point.go]: done

File: ecc/p384/point.go [hunk 3]
```diff
@@ -246,3 +268,95 @@ func (P *jacobianPoint) double() {
...
+func (P *homogeneousPoint) toAffine() *affinePoint {
+	var aP affinePoint
+	z := &fp384{}
+	fp384Inv(z, &P.z)
+	fp384Mul(&aP.x, &P.x, z)
+	fp384Mul(&aP.y, &P.y, z)
+	return &aP
+}
+
+// add calculates P=Q+R using complete addition formula for prime groups.
```
[the remark this answers is not in the document]
Comment (armfazh) [reply ecc/p384/point.go]: added

**Comment by henrydcase** [comment]
LGTM, Some small comments.

**Pull request merged**